File: ocaml/sat_decode.ml
(* Exhaustive encode/decode bijection. Same claim as sim/sat_decode.py. *)

let () =
  for word = 0 to 65535 do
    let op, side, delay, payload = Isa.decode word in
    let rebuilt = Isa.encode op payload delay side in
    if rebuilt <> word then
      failwith (Printf.sprintf "round-trip %04x -> %04x" word rebuilt);
    if op > 7 || delay > 15 || side > 1 then
      failwith "field range"
  done;
  for op = 0 to 7 do
    for delay = 0 to 15 do
      for side = 0 to 1 do
        for payload = 0 to 255 do
          let w = Isa.encode op payload delay side in
          let op', side', delay', payload' = Isa.decode w in
          if (op', side', delay', payload') <> (op, side, delay, payload) then
            failwith "encode mismatch"
        done
      done
    done
  done;
  print_endline "OCaml SAT-on-decode: 65536-word bijection OK"

File: sim/sat_decode.py
"""SAT-on-decode: encode/decode is a bijection on the 16-bit word.

A one-hot loop over bits[15:13] is a tautology. This file proves the
field split we claimed: op, side, delay, payload pack and unpack.
"""

from __future__ import annotations

from isa import OP_NAMES, decode_fields, encode


def prove_decode() -> None:
    for word in range(65536):
        f = decode_fields(word)
        rebuilt = encode(f["op"], f["payload"], f["delay"], f["side"])
        if rebuilt != word:
            raise AssertionError(
                f"round-trip {word:#06x} -> {f} -> {rebuilt:#06x}"
            )
        if f["op"] > 7 or f["delay"] > 15 or f["side"] > 1:
            raise AssertionError(f"field range {word:#06x} {f}")

    counts = [0] * 8
    for op in range(8):
        for delay in range(16):
            for side in range(2):
                for payload in range(256):
                    w = encode(op, payload, delay, side)
                    f = decode_fields(w)
                    if f != {"op": op, "side": side, "delay": delay, "payload": payload}:
                        raise AssertionError(f"encode mismatch {f}")
                    counts[op] += 1
    for i, n in enumerate(counts):
        if n != 8192:
            raise AssertionError(f"{OP_NAMES[i]} occupancy {n}")


if __name__ == "__main__":
    prove_decode()
    print("SAT-on-decode: 65536-word encode/decode bijection OK")
